protocols/request-response/src/json.rs

```diff
@@ -0,0 +1,195 @@
+///
+/// # Example
+///
+/// ```
+/// # use libp2p_request_response::{json, ProtocolSupport, self as request_response};
+/// # use libp2p_swarm::{StreamProtocol};
+/// #[derive(Debug, serde::Serialize, serde::Deserialize)]
+/// struct GreetRequest {
+///     name: String,
+/// }
+///
+/// #[derive(Debug, serde::Serialize, serde::Deserialize)]
+/// struct GreetResponse {
+///     message: String,
+/// }
+///
+/// let behaviour = json::Behaviour::<GreetRequest, GreetResponse>::new(
+///     [(StreamProtocol::new("/my-json-protocol"), ProtocolSupport::Full)],
+///     request_response::Config::default()
+/// );
+/// ```
+pub type Behaviour<Req, Resp> = crate::Behaviour<codec::Codec<Req, Resp>>;
+
+mod codec {
+    use crate::codec::read_to_end;
+    use async_trait::async_trait;
+    use futures::prelude::*;
+    use futures::{AsyncRead, AsyncWrite};
+    use libp2p_swarm::StreamProtocol;
+    use serde::{de::DeserializeOwned, Serialize};
+    use std::{io, marker::PhantomData};
+
+    /// Max request size in bytes
+    const REQUEST_SIZE_MAXIMUM: usize = 1024 * 1024;
+    /// Max response size in bytes
+    const RESPONSE_SIZE_MAXIMUM: usize = 10 * 1024 * 1024;
+
+    pub struct Codec<Req, Resp> {
+        phantom: PhantomData<(Req, Resp)>,
+    }
+
+    impl<Req, Resp> Default for Codec<Req, Resp> {
+        fn default() -> Self {
+            Codec {
+                phantom: PhantomData,
+            }
+        }
+    }
+
+    impl<Req, Resp> Clone for Codec<Req, Resp> {
+        fn clone(&self) -> Self {
+            Self::default()
+        }
+    }
+
+    #[async_trait]
+    impl<Req, Resp> crate::Codec for Codec<Req, Resp>
+    where
+        Req: Send + Serialize + DeserializeOwned,
+        Resp: Send + Serialize + DeserializeOwned,
+    {
+        type Protocol = StreamProtocol;
+        type Request = Req;
+        type Response = Resp;
+
+        async fn read_request<T>(&mut self, _: &Self::Protocol, io: &mut T) -> io::Result<Req>
+        where
+            T: AsyncRead + Unpin + Send,
+        {
+            let vec = read_to_end(io, REQUEST_SIZE_MAXIMUM).await?;
+
+            Ok(serde_json::from_slice(vec.as_slice())?)
+        }
+
+        async fn read_response<T>(&mut self, _: &Self::Protocol, io: &mut T) -> io::Result<Resp>
+        where
+            T: AsyncRead + Unpin + Send,
+        {
+            let vec = read_to_end(io, RESPONSE_SIZE_MAXIMUM).await?;
+
+            Ok(serde_json::from_slice(vec.as_slice())?)
+        }
+
+        async fn write_request<T>(
+            &mut self,
+            _: &Self::Protocol,
+            io: &mut T,
+            req: Self::Request,
+        ) -> io::Result<()>
+        where
+            T: AsyncWrite + Unpin + Send,
+        {
+            let data = serde_json::to_vec(&req)?;
+
+            io.write_all(data.as_ref()).await?;
+            io.flush().await?;
+            io.close().await?;
```

Comment:
Let's remove these, I am pretty sure the `request_response` code around this already does that.

Reply:
I can remove `io.flush().await?;`, but without `close` the test is hung.

Reply:
I fixed this comment